### PR TITLE
fix: Fixed the for_each value depends on resource attributes that cannot be determined until apply

### LINKED ISSUE #59: The "for_each" value depends on resource attributes that cannot be determined until apply
## Description

Hello. When using the latest module version and running the plan to create a _record_ based on the value of a CloudFront distribution created using a _for_each_: 

```tf
  source  = "terraform-aws-modules/route53/aws//modules/records"
  version = "2.3.0"

  …

  zone_id = local.dns_zone_id
  depends_on = [module.cloudfront]

  records = [
    {
      name = ""
      type = "A"
      alias = {
        name    = module.cloudfront.web.cloudfront_distribution_domain_name
        zone_id = module.cloudfront.web.cloudfront_distribution_hosted_zone_id
      }
    }
  ]
```

The plan fails with the following error when running locally or on Terraform Cloud:

```tf
Configuring remote state backend...
Initializing Terraform configuration...
╷
│ Error: Invalid for_each argument
│
│   on .terraform/modules/records/modules/records/main.tf line 32, in resource "aws_route53_record" "this":
│   32:   for_each = var.create && (var.zone_id != null || var.zone_name != null) ? local.recordsets : tomap({})
│     ├────────────────
│     │ local.recordsets will be known only after apply
│     │ var.create is true
│     │ var.zone_id is "ZLHLLCNNHTTH8"
│     │ var.zone_name is null
│
│ The "for_each" value depends on resource attributes that cannot be
│ determined until apply, so Terraform cannot predict how many instances will
│ be created. To work around this, use the -target argument to first apply
│ only the resources that the for_each depends on.
```

## Versions
```
Terraform v1.0.8
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.61.0
```

## Reproduction
- Are you using workspaces?: Yes
- Have you cleared the local cache: Yes and this error is even shows in TF Cloud.

### Code Snippet to Reproduce
Additionally to the above code, this is the extra relevant part:

`main.tf`
```tf
module "cloudfront" {
  source = "./modules/aws-cloudfront"

  for_each = {
    for k, v in var.projects : k => v
    if v.cdn
  }

  …
```
`varaibles.tf`
```tf
variable "projects" {
  type        = map(map(string))
  default = {
    "web" = {
      name = "…",
      cdn  = true,
    },
  }
}
```

## Expected behavior
The plan should end successfully.

## Actual behavior
The error is shown in the "plan" state and thus the apply is impossible.

### Terminal Output Screenshot(s)
image

## Workaround
On local backend, the `-target` argument can be used to first plan and apply the dependencies. On Terraform Cloud there is the `TF_CLI_ARGS_plan` environment variable, which is supported since [version 0.13](https://github.com/hashicorp/terraform/blob/v0.13/CHANGELOG.md), e.g.:
`TF_CLI_ARGS_plan="-target=module.cloudfront"`

## Additional context
When first creating the AWS CloudFront dependency, then the records can be planned correctly, as the resources now exists. But running per parts is not ideal.

The error is about "`local.recordsets will be known only after apply`". I wonder if it is possible to have this refactor to avoid the `for_each` issue, or if this is a deeper rooted Terraform limitation?
https://github.com/terraform-aws-modules/terraform-aws-route53/blob/533f07d83f3993a9ca191d200f49caba71114aa9/modules/records/main.tf#L15-L20

Thanks.

### PR DESCRIPTION
**Pull request by antonbabenko**
~It is still not working, and after wasting more than 20 hours on fixing #59 ... I give up (for now)...~

~If anyone likes challenges, please try to finish this PR.~

~All information and "definition of done" is [modules/records/main.tf](https://github.com/terraform-aws-modules/terraform-aws-route53/pull/72/files#diff-15c488fdc1301bb287fd6b94ed88c499fba27aeb9615824838ce468e96dc11a0).~

Huge thanks to @flora-five for figuring out the working solution.

Fixes #59

### CODE REVIEW

**Comment by flora-five** [comment]
When working on my previous PR, I didn't consider this:

  3. Terragrunt with "known values" - (TG always resolves values before calling Terraform). TG wraps values with `jsonencode()`.

Then the code in examples/complete that simulates Terragrunt is not a good simulation because it uses values not known yet.

With known values for the Terragrunt simulation, `var.records_jsonencoded` enabled and a small change to `local.all_records` (see https://github.com/flora-five/terraform-aws-route53/commit/c72578eb1255fc82f2805f985a1bcabc8ef3b1db), the plan is ok for Terraform versions 0.13.7 to 1.2.3. I didn't actually apply the example, just tested that the plan does not fail with the known issue.

**Comment by antonbabenko** [comment]
@flora-five Thanks for the update. I will try to find time later tonight to try to run an example with your commit.

I will make sure to provide static values (known values) for the example for Terragrunt.

File: modules/records/main.tf [hunk 1]
```diff
@@ -1,23 +1,10 @@
 locals {
-  # terragrunt users have to provide `records` as jsonencode()'d string.
+  # Terragrunt users have to provide `records_jsonencoded` as jsonencode()'d string.
   # See details: https://github.com/gruntwork-io/terragrunt/issues/1211
-  records = try(jsondecode(var.records_jsonencoded), var.records)
+  records = concat(var.records, try(jsondecode(var.records_jsonencoded), []))
```
Comment (antonbabenko) [inline modules/records/main.tf:4]: @flora-five THIS `concat` was the magic solution! Thank you very much for figuring it out!

**Comment by antonbabenko** [comment]
This PR is included in version 2.8.1 :tada:

**Pull request merged**